src/test/scala/org/kframework/minikore/NodeTests.scala

```diff
@@ -0,0 +1,205 @@
+
+    val intVar = b.Variable("A", Sort("Int"))
+
+    val e1: Pattern = b.And(intVar, int1)
+
+    val e2: Pattern = b.And(b.Variable("C", Sort("String")), stringFoo)
+
+    val plusApp: Pattern = b.Application(Symbol("Plus"), Seq(int1, int2))
+
+    val simpleVariable: Variable = b.Variable("X", Sort("Test"))
+
+    val simpleDomainValue: DomainValue = b.DomainValue(Symbol("test"), "String")
+
+    val top: Top = b.Top()
+
+    val bottom: Bottom = b.Bottom()
+
+    val simpleAnd: And = b.And(top, bottom)
+
+    val simpleOr: Or = b.Or(top, bottom)
+
+    val simpleNot: Not = b.Not(top)
+
+    val simpleImplies: Implies = b.Implies(bottom, top)
+
+    val simpleIntVar: Variable = b.Variable("A", Sort("Int"))
+
+    val simpleExists: Exists = b.Exists(simpleIntVar, b.Equals(simpleIntVar, int1))
+
+    val simpleForAll: ForAll = b.ForAll(simpleIntVar, b.Equals(simpleIntVar, b.Variable("Y", Sort("Int"))))
+
+    val simpleEquals: Equals = b.Equals(simpleOr, top)
+
+    val simpleList: Seq[Pattern] = Seq(top, bottom, simpleAnd, simpleOr, simpleNot, simpleImplies,
+      simpleIntVar, simpleExists, simpleForAll, simpleEquals, plusApp, e1, e2)
+  }
+
+
+  def b: build.Builders = DefaultBuilders
+
+  def t = TestPatterns
+
+  def identityF: (Pattern) => Pattern = (x) => x
+
+  @Test def sizeTest1(): Unit = {
+    assert(TestFunctions.size(TestPatterns.e1) == 2)
+  }
+
+  @Test def sizeTest2(): Unit = {
+    assert(TestFunctions.size(b.Equals(t.plusApp, t.int1)) == 3)
+  }
+
+  @Test def labelledNodeCountTest(): Unit = {
+    assert(TestFunctions.getLabelledNodesCount(t.b.Equals(t.intVar, t.plusApp)) == 1)
+  }
+
+  @Test def node1CountTest(): Unit = {
+    assert(TestFunctions.getNode1Count(t.simpleNot) == 1)
+  }
+
+  @Test def node2CountTest(): Unit = {
+    val n2Pattern: Pattern = b.And(t.simpleAnd, b.And(t.simpleAnd, b.Bottom()))
+    println(TestFunctions.getNode2Count(n2Pattern))
+
+    assert(TestFunctions.getNode2Count(n2Pattern) == 4)
+  }
+
+  @Test def node0CountTest(): Unit = {
+    val n0Pattern = b.And(b.Top, b.Bottom())
+
+    assert(TestFunctions.getNode0Count(n0Pattern) == 2)
+  }
+
+  @Test def leaf2CountTest(): Unit = {
+    val leafPattern = b.Equals(TestPatterns.simpleIntVar, TestPatterns.int1)
+
+    assert(TestFunctions.getLeaf2Count(leafPattern) == 2)
+  }
+
+  @Test def identityFunctionTest(): Unit = {
+    val pList: Seq[Pattern] = TestPatterns.simpleList
+
+    assert(pList.map(TestFunctions.map(identityF)).equals(pList))
+  }
+
+  @Test def simpleQuantifierTests(): Unit = {
+    def changeVar: (Pattern) => Pattern = {
+      case Variable(name, Sort(s)) => b.Variable("#" + name, Sort(s))
+      case n@_ => n
+    }
+
+    val changedVar: Variable = b.Variable("#A", Sort("Int"))
+
+    val changedExists: Exists = b.Exists(changedVar, b.Equals(changedVar, TestPatterns.int1))
+
+    assert(TestFunctions.map(changeVar)(TestPatterns.simpleExists) == changedExists)
+  }
+
```

Comment:
We have complete code coverage. The tests for unparsing cover all the ML constructs. Since I wrote NodeTest.scala to only cover Node behavior, I'll have to add a different set of tests that mixes both Node and Pattern behavior. I think that will be well covered in the next iteration, when we'll have utils + meta level, and tests for them.